src/server/game/Spells/Spell.cpp

```diff
@@ -5484,9 +5484,16 @@ SpellCastResult Spell::CheckCast(bool strict)
             case SPELL_EFFECT_TALENT_SPEC_SELECT:
             {
                 ChrSpecializationEntry const* spec = sChrSpecializationStore.LookupEntry(m_misc.SpecializationId);
-                if (!spec || spec->ClassID != m_caster->getClass())
+                Player* player = m_caster->ToPlayer();
+
+                if (!spec || (spec->ClassID != m_caster->getClass() && !spec->IsPetSpecialization()))
                     return SPELL_FAILED_NO_SPEC;
 
+                if (spec->IsPetSpecialization())
+                    if (Pet* pet = player->GetPet())
+                        if (!pet->IsPet() || ((Pet*)pet)->getPetType() != HUNTER_PET || pet->GetOwnerGUID() != player->GetGUID() || !pet->GetCharmInfo())
```

Comment:
pet is always a pet, `pet->IsPet()` is useless
casting pet to pet is also useless `(Pet*)pet`
you also need to return SPELL_FAILED_NO_PET if !pet
`pet->GetOwnerGUID() != player->GetGUID()` is it even possible to have a not owned pet?